packages/react/src/runtime/style-cache.tsx

```diff
@@ -1,24 +1,70 @@
 import * as React from 'react';
 import { createContext, useContext } from 'react';
 import { isNodeEnvironment } from './is-node';
+import insertRule from './sheet';
 import { ProviderComponent, UseCacheHook } from './types';
 
 /**
  * Cache to hold already used styles.
  * React Context on the server - singleton object on the client.
  */
-const Cache: any = isNodeEnvironment() ? createContext<Record<string, true> | null>(null) : {};
+let Cache: any = isNodeEnvironment() ? createContext<Record<string, true> | null>(null) : {};
 
 if (!isNodeEnvironment()) {
+  const ssrCache = document.querySelector('script[data-cmpld="c"]');
```

Comment:
I ran a bunch of performance tests on the other method and it started to get quite bad when there were ~1000 SSR'd style tags on the DOM. This creates the cache immediately. Then we can move all the SSR'd style tags to the head where we can asynchronously begin to consolidate them using ```insertRule```.